### PR TITLE
allwinner: Add BL32(corresponds to Trusted OS) support

### PR DESCRIPTION
**Pull request by Amit-Radur**
This patch is an attempt to run Trusted OS(OP-TEE OS being one of them) along
side BL31 image.

ATF supports multiple SPD's that can take dispatcher name(opteed for OP-TEE OS)
as an input using the 'SPD=<dispatcher name>' option during bl31 build.

Signed-off-by: Amit Singh Tomar <amittomer25@gmail.com>

### CODE REVIEW

File: plat/allwinner/common/include/platform_def.h [hunk 1]
```diff
@@ -39,4 +39,13 @@
 #define PLATFORM_MMAP_REGIONS		4
 #define PLATFORM_STACK_SIZE		(0x1000 / PLATFORM_CORE_COUNT)
 
+#define PLAT_ARM_NS_IMAGE_OFFSET    0x4a000000
+
+#ifdef SPD_opteed
+#define BL32_BASE                   SUNXI_DRAM_BASE
```
Comment (Andre-ARM) [inline plat/allwinner/common/include/platform_def.h:45]: Is there any reason this needs to be set here? I would suggest:
```
#ifndef SPD_none
#ifndef BL32_BASE
#define BL32_BASE        SUNXI_DRAM_BASE
#endif
#endif
```

File: plat/allwinner/common/sunxi_bl31_setup.c [hunk 2]
```diff
@@ -12,11 +12,13 @@
 #include <gicv2.h>
 #include <platform.h>
 #include <platform_def.h>
+#include <plat_arm.h>
```
Comment (Andre-ARM) [inline plat/allwinner/common/sunxi_bl31_setup.c:15]: I don't think this is needed.

File: plat/allwinner/common/sunxi_bl31_setup.c [hunk 3]
```diff
@@ -32,16 +34,26 @@ void bl31_early_platform_setup(bl31_params_t *from_bl2,
...
 
+#ifdef BL32_BASE
+	/* Populate entry point information for BL32 */
+	SET_PARAM_HEAD(&bl32_image_ep_info, PARAM_EP, VERSION_1, 0);
+	SET_SECURITY_STATE(bl32_image_ep_info.h.attr, SECURE);
+	/* Tell BL32 where the trusted software image
+	 * is located and the entry state information
+	 */
+	bl32_image_ep_info.pc = BL32_BASE;
+	bl32_image_ep_info.spsr = arm_get_spsr_for_bl32_entry();
```
Comment (Andre-ARM) [inline plat/allwinner/common/sunxi_bl31_setup.c:45]: According to the comment in the definition of this function this is not needed: https://github.com/Andre-ARM/arm-trusted-firmware/blob/f11916bf1dfecd7ee76240eebd847681ce366a67/plat/arm/common/arm_common.c#L108-L118

File: plat/allwinner/common/sunxi_bl31_setup.c [hunk 4]
```diff
@@ -32,16 +34,26 @@ void bl31_early_platform_setup(bl31_params_t *from_bl2,
...
+#endif
 	/* Populate entry point information for BL33 */
 	SET_PARAM_HEAD(&bl33_image_ep_info, PARAM_EP, VERSION_1, 0);
+	SET_SECURITY_STATE(bl33_image_ep_info.h.attr, NON_SECURE);
```
Comment (Andre-ARM) [inline plat/allwinner/common/sunxi_bl31_setup.c:49]: Please don't move this around needlessly.

File: plat/allwinner/common/sunxi_bl31_setup.c [hunk 5]
```diff
@@ -68,7 +80,12 @@ void bl31_platform_setup(void)
 entry_point_info_t *bl31_plat_get_next_image_ep_info(uint32_t type)
 {
 	assert(sec_state_is_valid(type));
-	assert(type == NON_SECURE);
 
-	return &bl33_image_ep_info;
+	if (type == NON_SECURE)
+		return &bl33_image_ep_info;
+
+	else if (type == SECURE && bl32_image_ep_info.pc)
```
Comment (Andre-ARM) [inline plat/allwinner/common/sunxi_bl31_setup.c:87]: The "else" is not needed.

File: plat/allwinner/sun50i_a64/platform.mk [hunk 6]
```diff
@@ -14,6 +14,7 @@ PLAT_INCLUDES		:=	-Iinclude/plat/arm/common/		\
 
 PLAT_BL_COMMON_SOURCES	:=	drivers/console/${ARCH}/console.S	\
 				drivers/ti/uart/${ARCH}/16550_console.S	\
+				plat/arm/common/arm_common.c \
```
Comment (Andre-ARM) [inline plat/allwinner/sun50i_a64/platform.mk:17]: This looks wrong, also is not needed anymore.

File: plat/allwinner/common/include/platform_def.h [hunk 7]
```diff
@@ -39,4 +39,13 @@
 #define PLATFORM_MMAP_REGIONS		4
 #define PLATFORM_STACK_SIZE		(0x1000 / PLATFORM_CORE_COUNT)
 
+#define PLAT_ARM_NS_IMAGE_OFFSET    0x4a000000
```
Comment (Andre-ARM) [inline plat/allwinner/common/include/platform_def.h:42]: Please indent the value with tabs to align with the other definitions.

**Review by Andre-ARM — CHANGES_REQUESTED** [review]
- So who does protect the memory for the secure OS? Is that ATF's duty or is this meant to be done by OPT-EE OS, for instance?
- This will probably fail horribly if there is nothing at this address. I take it it's the duty of the loader (U-Boot's SPL in our case) to put some sensible code there, but what if this didn't happen? If possible, I would like to keep this one-ATF-image approach, so can we somehow check that there has been something loaded into DRAM? Maybe OPTEE-OS has some magic we can check?

File: plat/allwinner/common/sunxi_bl31_setup.c (hunk 3 above)
Comment (Amit-Radur) [reply plat/allwinner/common/sunxi_bl31_setup.c:45]: Ok, you mean we need not to fill bl32_image_ep_info.spsr via arm_get_spsr_for_bl32_entry ?

File: plat/allwinner/common/include/platform_def.h (hunk 7 above)
Comment (Amit-Radur) [reply plat/allwinner/common/include/platform_def.h:42]: ok.

File: plat/allwinner/common/include/platform_def.h (hunk 1 above)
Comment (Amit-Radur) [reply plat/allwinner/common/include/platform_def.h:45]: Wanted to set BL32_BASE only in case of SPD has been used.

**Comment by Amit-Radur** [comment]
So who does protect the memory for the secure OS? Is that ATF's duty or is this meant to be done by OPT-EE OS, for instance?
 
[Amit]: AFAIK, ideally, it should be done by the loader(could be running SPL or ATF running at EL3), prior to OPTEE-OS

Thought about it, OPTEE-OS does provide driver both TZC400 and TZC380 but wondering can it
be programmed from Secure EL1 for armv8 platform?

 This will probably fail horribly if there is nothing at this address. I take it it's the duty of the loader (U-Boot's SPL in our case) to put some sensible code there, but what if this didn't happen? If possible, I would like to keep this one-ATF-image approach, so can we somehow check that there has been something loaded into DRAM? Maybe OPTEE-OS has some magic we can check?

[Amit]: Yes, OP-TEE OS header has the magic number. But what if someone prefers to load raw arm executable provided by the OPTEE-OS build system(tee-pager_v2.bin, it doesn't have the magic number)?

Also, this looks to be a bit more involved(at least for me), parsing optee header support is already 
present in ATF(https://github.com/ARM-software/arm-trusted-firmware/blob/master/lib/optee/optee_utils.c#L131) but it needs enabling  "LOAD_IMAGE_V2" type for allwinner platform.

After hacking "LOAD_IMAGE_V2"  thing, could read optee header but optee image parsing failed at
(https://github.com/ARM-software/arm-trusted-firmware/blob/master/lib/optee/optee_utils.c#L101), not 
sure we need all that at this moment?

Right now,  tested optee-os port with tee-pager_v2.bin that is just a plain arm executable.

There are few more images generated by OPTEE-OS build system:

1). tee-header_v2.bin, image just having a header:

 hexdump tee-header_v2.bin 
0000000 504f 4554 0102 0000 0001 0000 0000 0000
0000010 0000 4000 0000 0000 8220 0004          
000001c

First, 4 bytes is OPTEE magic number "OPTE", combined it with tee-pager_v2.bin
    # cat tee-header_v2.bin tee-pager_v2.bin > opt.bin 

and then roughly tried this to check if there is something in DRAM at 0x40000000

https://paste.ubuntu.com/p/58c34q6bVP/

This check worked but OPTEE-OS failed to boot(stuck at):

INFO:    PSCI: Powering off cluster 0 core 1
INFO:    PSCI: Disabling power to cluster 0 core 1
INFO:    PSCI: Powering off cluster 0 core 2
INFO:    PSCI: Disabling power to cluster 0 core 2
INFO:    PSCI: Powering off cluster 0 core 3
INFO:    PSCI: Disabling power to cluster 0 core 3
NOTICE:  BL31: v1.4(debug):v1.3-1318-g2dae6e6-dirty
NOTICE:  BL31: Built : 23:06:04, May 27 2018
INFO:    ARM GICv2 driver initialized
INFO:    BL31: Platform setup done
INFO:    BL31: Initializing runtime services
INFO:    BL31: cortex_a53: errata workaround for 835769 was applied
INFO:    BL31: cortex_a53: errata workaround for 843419 was applied
INFO:    BL31: cortex_a53: errata workaround for 855873 was applied
INFO:    BL31: Initializing BL32

Wondering, what is going on :(

**Comment by Andre-ARM** [comment]
* Re: OPTEE-OS header parsing:
I don't think that's the right approach. The "OTPE" header is just meant for the *loader*, so U-Boot's SPL in our case. This loader is meant to parse the header and load the raw binary to the location specified in this very header. So the magic number is not there anymore in memory. You can easily see that this can't work since the header starts with the magic and this is not a valid AArch64 instruction.
Can you check what Rockchip does in U-Boot/ATF to communicate loaded payloads? I believe they have something in place to let ATF know what the loaded did.
* Re: Secure memory setup:
Since ATF can't be bothered with knowing how much memory the secure payload is going to need, I guess it's up to BL32/OPTEE-OS to secure the memory. The TZC is memory mapped, and since BL32 is also running secure, I can't see why programming it should not work. Just need to figure out how the two (BL31 and OPTEE-OS) would coordinate over this. Again looking at existing examples (Juno? Rockchip?) would probably be helpful.

**Comment by Amit-Radur** [comment]
Re:Secure memory setup:
You're right about it. Now, could program TZC/SMC from OPTEE-OS itself running in Secure EL1.
But what about the SPC block, it anyways needed to be programmed in ATF and should we re-progarm it in OPTEE-OS ?

"Just need to figure out how the two (BL31 and OPTEE-OS) would coordinate over this"
I didn't get why it would be needed ?

**Pull request closed without merging**